# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
# Установим, что наличие задолженности(задержки) по кредиту в нашем датафрейме определяется значением "1" в соответствующем столбце "debt" (см. задание №2)
# Тогда нужно создать сводную таблицу, в которой для каждой категории количества детей отображается общее количество записей и процент случаев задолженности по кредиту ('percentage').

# Задаем значения для таблицы
table_children = data.pivot_table(index='children', values='debt', aggfunc=['count', 'mean'])
# Задаем названия её столбцов
table_children.columns = ['total_count', 'percentage']

# Выводим таблицу на экран
print(table_children)

          total_count  percentage
children                         
0               14091    0.075438
1                4808    0.092346
2                2052    0.094542
3                 330    0.081818
4                  41    0.097561
5                   9    0.000000


<div class="alert alert-success">

**Артём:** 👍 Молодец, что используешь сводную таблицу. 

</div>

**Вывод:**

Как можно заметить, в сводной таблице минимальному значению в столбце percentage соответствует 6 строка таблицы, т.е при минимальной выборке из 9 человек с количеством детей равным 5, задолженностей по кредитам нет ни у кого.
Но так как выборка несущественна, этот пункт в списке значений по возрастанию процента должников (percentage) можно просто игнорировать.

На первом месте по возрастанию в столбце percentage идет 1 строка таблицы (0), т.е при самой объемной выборке в 14091 человека с количеством детей, равным 0 (без детей вообще) - чаще всего задолженностей по кредиту не имеется.

Остальные же случаи c количеством детей больше 0 имеют больший процент задолженностей по кредиту, причем значения столбца "percentage" лишь увеличивается с увеличением количества детей (с исключением лишь при количестве детей равном 3), а значит можно сделать вывод, что есть прямая корреляция между количеством детей и возвратом кредита в срок

<div class="alert alert-success">

**Артём:** 👍 Отлично.

</div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [23]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
# С этим заданием можно справиться аналогичным с предыдущим способом:
table_family_status = data.pivot_table(index='family_status', values='debt', aggfunc=['count', 'mean'])
table_family_status.columns = ['total_count', 'percentage']
table_family_status_sorted = table_family_status.sort_values(by='total_count', ascending=False)

# Рассчитываем процентное соотношение клиентов с разным семейным положением:
total = table_family_status_sorted['total_count'].sum()
table_family_status_sorted['percentages'] = (table_family_status_sorted['total_count'] / total) * 100

print(table_family_status_sorted)

                       total_count  percentage  percentages
family_status                                              
женат / замужем              12261    0.075606    57.479724
гражданский брак              4134    0.093130    19.380245
Не женат / не замужем         2796    0.097639    13.107684
в разводе                     1189    0.070648     5.574047
вдовец / вдова                 951    0.066246     4.458300


**Вывод:** 

При семейном положении "женат/замужем", которое составляет 57% выборки(столбец "percentages"), процент задолженности составляет 7.56%, категории "в разводе" и "вдовец / вдова" имеют более низкий процент задолженности (7.06% и 6.62% соответственно), что может свидетельствовать о более низком риске невозврата для этих категорий клиентов, категории "гражданский брак" и "Не женат / не замужем" имеют более высокий процент задолженности (9.31% и 9.76% соответственно), что может указывать на более высокий риск невозврата кредита для людей в этих категориях семейного положения.

Таким образом, можно сказать, что существует определенная зависимость между семейным положением и возвратом кредита в срок

<div class="alert alert-success">

**Артём:** 👍 Верно.

</div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [24]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
sorted_data = data.sort_values(by='total_income')


# Создадим 2 категории дохода: высокий(больше 150к) и низкий(до 150к) (условно, так как объем выборки категорий примерно равен при таком распределении)
filtered_data_greater_than_200k = sorted_data[sorted_data['total_income'] > 150000]

filtered_data_less_than_200k = sorted_data[sorted_data['total_income'] < 150000]

data['income_category'] = pd.cut(data['total_income'], bins=[-float('inf'), 150000, float('inf')], labels=['Low', 'High'], right=False)

# Подсчет процента задолженностей в каждой категории
table_income = data.pivot_table(index='income_category', values='debt', aggfunc=['count', 'mean'])
table_income.columns = ['total_count', 'default_percentage']

# Вывод таблицы
print(table_income)

                 total_count  default_percentage
income_category                                 
Low                    11554            0.084213
High                    9777            0.077631


**Вывод:** 

Клиенты с более высоким доходом чаще платят кредит в срок (7.7% должников по сравнению с 8.4%)

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [25]:
table_purpose = data.pivot_table(index='purpose_category', values='debt', aggfunc=['count', 'mean'])
table_purpose.columns = ['total_count', 'default_percentage']
print(table_purpose)

                          total_count  default_percentage
purpose_category                                         
операции с автомобилем           4279            0.093480
операции с недвижимостью        10751            0.072551
получение образования            3988            0.092528
проведение свадьбы               2313            0.079118


**Вывод:** 

Заемщики, укажавшие цель "операции с недвижимостью" и "проведение свадьбы", имеют более низкий процент задолженности (примерно 7.26% и 7.91%). Заемщики, указавшие цель "операции с автомобилем" и "получение образования", имеют самый высокий процент задолженности (около 9.35% и 9.25%). Таким образом, можно сказать, что цель кредита может влиять на его возврат в срок

<div class="alert alert-success">

**Артём:** 👍 Правильно.

</div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

Неправильное заполнение данными, некорректные исходные данные, остутствие данных

<div class="alert alert-success">

**Артём:** 👍 Согласен.

</div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

Когда у нас есть пропуски в числовых значениях, заполнение их медианным значением - это хороший выбор, так как:

1. Медиана не так чувствительна к необычным значениям (выбросам), что делает ее более устойчивой
2. Этот метод помогает сохранить общую форму (распределение) наших данных, что важно для правильного понимания ситуации

Поэтому заполнение медианным значением - это практичный метод, который часто используется

<div class="alert alert-success">

**Артём:** 👍 Хорошо.

</div>

### Шаг 4: общий вывод.

Разные факторы могут существенно влиять на погашение кредита в срок, а именно:

1. Влияние количества детей на возврат кредита:
Заемщики с детьми имеют более высокий процент задолженности по сравнению с бездетными заемщиками, и в среднем вероятность возникновения задолженности возрастает с увеличением количества детей.

2. Влияние семейного положения на возврат кредита:
Заемщики, находящиеся в гражданском браке, имеют самый высокий процент задолженности по сравнению с остальными категориями семейного статуса. Вдовцы/вдовы, разведенные, и женатые/замужем люди, напротив, демонстрируют наименьший процент задолженности.

3. Влияние уровня дохода на возврат кредита:
Более высокий уровень дохода существенно связан с более низким процентом задолженности. Заемщики с высоким уровнем дохода имеют значительно более низкий процент задолженности, чем те, у кого доход ниже.

4. Влияние цели кредита на возврат в срок:
Заемщики, берущие кредит на операции с авто, демонстрируют более высокий процент задолженности по сравнению с теми, кто берет кредит на операции с недвижимостью. Похожие тенденции видны и для целей получения образования и проведения свадьбы.

Вывод:
- Различные факторы, такие как семейное положение, наличие детей, уровень дохода и цель кредита, оказывают влияние на возвратность кредита в срок. При принятии решения о выдаче кредита стоит учитывать не только формальные критерии, но и дополнительные факторы, чтобы сделать более обоснованные и риско-ориентированные решения.
- Полученные результаты могут использоваться для оптимизации кредитного скоринга и более эффективного управления рисками в банковской сфере.

<div class="alert alert-success">

**Артём:** 👍 Хороший вывод. Аккуратно, последовательно приятно читать.

</div>